torch/csrc/autograd/profiler.cpp

```diff
@@ -156,98 +281,126 @@ void enableProfiler(ProfilerConfig config) {
-                "Did not find thread_id matching ",
-                fn.getStartCallbacksThreadId());
-
-            auto& eventList = eventListIter->second;
-            eventList->record(
-                      EventKind::PopRange,
-                      StringView(""),
-                      fn.getStartCallbacksThreadId(),
-                      state == ProfilerState::CUDA);
-          }
-        } else {
-          popRange();
+        const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+        auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+        if (!state_ptr || state_ptr->config().state == ProfilerState::Disabled) {
+          return;
         }
+        state_ptr->popRange(fn.getStartCallbacksThreadId());
       },
-      /* needs_inputs */ config.report_input_shapes,
-      /* sampling_prob */ 1.0,
-      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE});
-  state = new_state;
-  c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, true);
+      needs_inputs,
+      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE},
+      CallbackKind::PROFILER);
+}
+
+void removeProfilingCallbacks() {
+  removeCallback(CallbackKind::PROFILER);
+}
 
-  if(state == ProfilerState::CUDA) {
+bool unused_ = []() {
+  at::ThreadLocalState::registerThreadLocalSetting(
+    at::ThreadLocalSetting::PROFILER,
+    []() {
+      auto v = at::SettingValue();
+      v.value = (int64_t)(profiler_nested_depth_ > 0);
+      return v;
+    },
+    [](at::SettingValue v) {
+      // push profiling callbacks in the child task
+      // if profiling is enabled in the parent
+      auto to_push = (bool)v.value;
+      if (to_push) {
+        if (profiler_nested_depth_ == 0) {
+          pushProfilingCallbacks();
+        }
+        profiler_nested_depth_++;
+      }
+      if (!to_push) {
+        --profiler_nested_depth_;
+        if (profiler_nested_depth_ == 0) {
+          removeProfilingCallbacks();
+        }
+      }
+    }
+  );
+  return true;
+}();
+
+const int kCUDAWarmupStart = 5;
+
+} // namespace
+
+void registerCUDAMethods(CUDAStubs* stubs) {
+  cuda_stubs = stubs;
+}
+
+ProfilerConfig::~ProfilerConfig() = default;
+
+bool profilerEnabled() {
+  const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+  auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+  return state_ptr && state_ptr->config().state != ProfilerState::Disabled;
+}
+
+void enableProfiler(const ProfilerConfig& new_config) {
+  TORCH_CHECK(new_config.state != ProfilerState::NVTX || cuda_stubs->enabled(),
+    "Can't use NVTX profiler - PyTorch was compiled without CUDA");
+
+  auto state = std::make_shared<ProfilerThreadLocalState>(new_config);
+  at::ThreadLocalDebugInfo::push(at::DebugInfoKind::PROFILER_STATE, state);
+
+  if (profiler_nested_depth_ == 0) {
+    pushProfilingCallbacks(new_config.report_input_shapes);
```

Comment:
Should we move the profiling callbacks into `ProfilerThreadLocalState`? It seems appropriate that each time we move to a different thread, we should carry over `ProfilerThreadLocalState` with the appropriate callbacks without having to call `pushProfilingCallbacks` each time?

Reply:
ok, but who will call these callbacks?
the only thing that right now that is capable of actually calling the hooks for each operator/method/user scope/etc is record function

I guess, the more interesting question is can we put all thread local callbacks not into a thread local vector in record_function.cpp but in something similar to DebugInfoCase struct, pointer to which is automatically propagated; but I think this is potentially in the scope of a separate PR

Reply:
In terms of calling the callbacks, record function and look up the thread local `ProfilerThreadLocalState` and call a method `runCallbacks` on that object? Something like:

```
at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE)->runCallbacks();
```

> I guess, the more interesting question is can we put all thread local callbacks (not only profiler but all thread local callbacks in record function) not into a thread local vector in record_function.cpp but in something similar to DebugInfoCase struct, pointer to which is automatically propagated; but I think this is potentially in the scope of a separate PR

Yes, this is similar to what I was thinking and maybe we can do this in a separate PR as you mentioned. I feel this would be a much cleaner design and we wouldn't need profiler_nested_depth and calls to explicitly removing callbacks.

Reply:
Agree that ideally we don't need this nested counter, I'm still not sure about the specific implementation. As Dima mentioned above, there's an improvement to ThreadLocalState we should do - make it possible to handle enter/exit style of settings, I think this might cover automatic pushing and popping of callbacks, I will move the change to ThreadLocalState into a separate PR and we will continue dicussion there; I think we can get rid of nested counter

Reply:
one hard design constraint is that RecordFunction should be as fast as possible, rn it just uses a thread local vector of callbacks

Reply:
originally what we're trying to do here is simple - avoid double adding of profiling callbacks;
one possibility is to extend RecordFunction a bit to account for the case of push/pop the same callback potentially in nested ranges

Reply:
@pritamdamania87 please check https://github.com/pytorch/pytorch/pull/37745
by making TLS callbacks automatically propagate-able, and disabling nested profiling we can cover TLS profiler use case without adding much code